### DISCUSSION TITLE
tailwind not being applied to library

### DISCUSSION THREAD

**Discussion by k3RetailDaveIzon** [tailwindlabs/tailwindcss#8002] [Help]
Hi , I have created a library in angular which is styled using tailwind. This is then been push to NPM and then imported into a new project, but the css is not getting applied. I have referenced the node-module path in my tailwind.config.ts:
```
content: [
    "./src/**/*.{html,ts}",
    './node_modules/components-name/**/*.{html,js,ts}'
  ],

```
What am i missing?

Tailwind is working if i apply it directly to the new application, it just doesn't work with the imported library.

**Comment by KoenVerhulst** [comment]
Hi,

I'm having the same problem.

I do think this is currently not possible. By adding the "node-modules/x" to the tailwind content config, I think you are justing scanning the files in order to see any used tailwind classes. As soon as you use @apply or other specific tailwind functionality inside the library, things stop working.

**Reply by k3RetailDaveIzon** [reply]
I found the answer to this:
You need to run:
npx tailwindcss -o ./src/lib/styles/tailwind.scss --minify

To create the css file that you then reference from the component:
e.g.

 styleUrls: ['../../styles/tailwind.scss']
 
 Then when you 'ng build', everything will be packaged together and work when you install the library.
 
 Hope this helps

**Comment by CharlyRipp** [comment]
@k3RetailDaveIzon Your solution would include all component's styles into all components (duplicate)?

But you inspired me to make a small process to do just this but per-component, only at package/publish time (never committed to the repo). I updated a stackoverflow answer with the goods: https://stackoverflow.com/a/72208906/9404093

As best practice- also putting the meat and potatoes here incase the answer is changed/lost:

Created a small node script to find each component and resolve it's tailwind `@apply` styles individually.

```typescript
import { readFile, writeFile } from "fs";
import { sync } from 'glob';
import { exec } from 'child_process';

const libRoot      = 'projects/my-lib/src/lib';
const tailwindConf = 'tailwind.config.js'; // may be apps/demo when using NX
const processedExt = '.precompiled.scss';
const styleRegex   = /styleUrls:\s*\[([^\]]+)]/;

// Find all `.scss` files and tailwind process them
sync(`${libRoot}/**/*.component.scss`).forEach(file => {
  const cssFile = file.replace(/\.scss$/, processedExt);
  exec(`npx tailwind -c ${tailwindConf} -i ${file} -o ${cssFile}`, (err, stdout, stderr) => {
    if (err) {
      console.error(stderr);
      throw err;
    }
  });
});

// .component.ts update
// Find all components with `styleUrls` and switch `.scss` extension to our precompiled file names
sync(`${libRoot}/**/*.component.ts`).forEach(file => {
  readFile(file, (err, data) => {
    if (err) throw err;
    const content = data.toString();
    const match = content.match(styleRegex);
    if (match) {
      const styleUrls = match[1]
        .split(',')
        .map(s => s.trim().replace('.scss', processedExt))
        .join(', ');

      writeFile(file, content.replace(styleRegex, `styleUrls: [${styleUrls}]`), (err) => {
        if (err) throw err;
      });
    }
  });
});
```

> This should only be ran by your CI process and never committed.
> Also this could easily be switched to javascript instead of typescript

Other possible ways to do this (untested) without the `.component.ts` update:
* Utilize `environment.prod.ts`'s `production: true` flag to decide the style file to use
    * `styleUrls: [ environment.prod ? 'my.component.precompiled.scss' : 'my.component.scss' ],`
    * Gotta remember this for all new components
* Change the tailwind compile to output to the same scss file
    * Less moving parts - I liked the separate file so I'd realize quickly if it were accidentally ran/committed

Then a very rough usage in your `package.json` for a CI build:

`"build:ci": "node --require ts-node/register projects/my-lib/src/precompile.ts && npm run build:my-lib"`
